scapy/contrib/ethercat.py

```diff
@@ -0,0 +1,647 @@
+        """
+
+        if type(s) is tuple and len(s) == 3:
+            s, bits_in_set, fields = s
+        else:
+            bits_in_set = 0
+            fields = []
+
+        bits_in_set += self.size
+
+        fields.append(self)
+
+        if bits_in_set % 8:
+            # we are in between the bitfield
+            return (s, bits_in_set, fields), None
+
+        else:
+            cur_val = 0
+            cur_val_bit_idx = 0
+            this_val = 0
+
+            field_idx = 0
+            field = fields[field_idx]
+            field_required_bits = field.size
+            idx = 0
+
+            s = bytearray(s)
+            bf_total_byte_length = bits_in_set // 8
+
+            for octet in s[0:bf_total_byte_length]:
+                idx += 1
+
+                octet_bits_left = 8
+
+                while octet_bits_left:
+
+                    if field_required_bits == octet_bits_left:
+                        # whole field fits into remaining bits
+                        # as this also signals byte-alignment this should exit the inner and outer loop
+                        cur_val |= octet << cur_val_bit_idx
+                        pkt.fields[field.name] = cur_val
+
+                        '''
+                        TODO: check if do_dessect() needs a non-None check for assignment to raw_packet_cache_fields
+
+                        setfieldval() is evil as it sets raw_packet_cache_fields to None - but this attribute
+                        is accessed in do_dissect() without checking for None... exception is catched and the
+                        user ends up with a layer decoded as raw...
+
+                        pkt.setfieldval(field.name, int(bit_str[:field.size], 2))
+                        '''
+
+                        octet_bits_left = 0
+
+                        this_val = cur_val
+
+                    elif field_required_bits < octet_bits_left:
+                        # pick required bits
+                        cur_val |= (octet & ((2 ** field_required_bits) - 1)) << cur_val_bit_idx
+                        pkt.fields[field.name] = cur_val
+
+                        # remove consumed bits
+                        octet >>= field_required_bits
+                        octet_bits_left -= field_required_bits
+
+                        # and move to the next field
+                        field_idx += 1
+                        field = fields[field_idx]
+                        field_required_bits = field.size
+                        cur_val_bit_idx = 0
+                        cur_val = 0
+
+                    elif field_required_bits > octet_bits_left:
+                        # take remaining bits
+                        cur_val |= octet << cur_val_bit_idx
+
+                        cur_val_bit_idx += octet_bits_left
+                        field_required_bits -= octet_bits_left
+                        octet_bits_left = 0
+
+            return s[bf_total_byte_length:], this_val
```

Comment:
Btw, small question: are those LEBitField universal of ethercat specific ?

We needed one of those in layers/bluetooth.py
https://github.com/secdev/scapy/blob/master/scapy/layers/bluetooth.py#L92
And instead have a predissect hack